### PR TITLE
Add Patient and Risk field

### PR DESCRIPTION
**Pull request by jiayi1129**
- Add `Patient` and `Risk` Classes
- `Patient` class is inherited from `Person` class as a `Patient` is a `Person`
- Add `PatientTest` and `RiskTest` to tests
- Add `PatientBuilder` and `TypicalPatients` to `testutils`

Read `Add Patient functionality to program` commit body below to understand some refactoring changes rationale

**Risk functionality**
[/rRISK]
RISK is either LOW, MEDIUM or HIGH
if field is left empty by user input, then risk is stored as `RISKLEVEL.UNCLASSIFIED` and written to Json as "UNCLASSIFIED".
*Added functionality: uppercase Risk such that r/low, r/Low works the same as r/LOW

### CODE REVIEW

**Review by Haozeee — COMMENTED** [review]
Overall, I think the code looks good with some minor changes needed here and there. Since there are a lot of changes, there may be some things that I missed out on. I'll try to take a look again when I have time.

While the code works for now, the OOP design of it can be improved. Since, we will be adding doctors in the future and patients and doctors will share a lot of similar fields and functionalities, we can abstract out some of the common parts in the OOP design. For example, we can have a abstract AddPersonCommand and then an abstract AddPatientCommand extending from it. The functionality related to the data fields in Person can be implemented in AddPersonCommand while the extra patient fields (Risk) can be added in the AddPatientCommand. This will allow us to have less repeated code when we add Doctors, and make the codebase easier to maintain as well.

File: src/main/java/seedu/plannermd/model/patient/Patient.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        if (!(other instanceof Patient)) {
+            return false;
+        }
+
+        Patient otherPerson = (Patient) other;
```
[the remark this answers is not in the document]
Comment (jiayi1129) [reply src/main/java/seedu/plannermd/model/patient/Patient.java:48]: Good catch! Will change thanks 👍

File: src/main/java/seedu/plannermd/model/patient/Patient.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        return otherPerson.getName().equals(getName())
+                && otherPerson.getPhone().equals(getPhone())
+                && otherPerson.getEmail().equals(getEmail())
+                && otherPerson.getAddress().equals(getAddress())
+                && otherPerson.getTags().equals(getTags())
+                && otherPerson.getRisk().equals(getRisk());
```
[the remark this answers is not in the document]
Comment (jiayi1129) [reply src/main/java/seedu/plannermd/model/patient/Patient.java:54]: Yes that's a good suggestion, I will make the changes as well.

File: src/main/java/seedu/plannermd/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -121,4 +122,19 @@ public static Tag parseTag(String tag) throws ParseException {
...
+
+    /**
+     * Parses a {@code String risk} into an {@code Risk}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code risk} is invalid.
+     */
+    public static Risk parseRisk(String risk) throws ParseException {
+        requireNonNull(risk);
+        String trimmedRisk = risk.trim();
```
[the remark this answers is not in the document]
Comment (jiayi1129) [reply src/main/java/seedu/plannermd/logic/parser/ParserUtil.java:134]: Sure, that sounds like a good to have and an easy addition

File: src/main/java/seedu/plannermd/storage/JsonAdaptedPatient.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import seedu.plannermd.model.patient.Risk;
+import seedu.plannermd.model.person.Address;
+import seedu.plannermd.model.person.Email;
+import seedu.plannermd.model.person.Name;
+import seedu.plannermd.model.person.Person;
+import seedu.plannermd.model.person.Phone;
+import seedu.plannermd.model.tag.Tag;
+
+/**
+ * Jackson-friendly version of {@link Person}.
```
[the remark this answers is not in the document]
Comment (jiayi1129) [reply src/main/java/seedu/plannermd/storage/JsonAdaptedPatient.java:23]: Good catch!

**Comment by jiayi1129** [comment]
Noting some discussions that I had with @Haozeee 

**Commands**
Perhaps it is possible to have a more general method named `AddPersonCommand` that inherits from `AddCommand`, then `AddPatientCommand` and `AddDoctorCommand` can inherit from the `AddPersonCommand`.

**Parsers**
Similar to _Commands_, perhaps it is possible to add a `AddPersonCommandParser` to bridge `AddDoctorCommandParser` / `AddPatientCommandParser` and `AddCommand` to simplify some common functions.

**UniquePersonList**
Perhaps it is possible to use this for both the unique patient and doctor list. However some methods like `public void setPersons(List<Person> persons)` and `public ObservableList<Person> asUnmodifiableObservableList()` may prove to be difficult to modify based on the invariant property of generics. I do believe changing the class to 
`public class UniquePersonList<T extends Person> implements Iterable<Person>` and then modifying the function's parameters would be a good way to move forward but there could be some dangerous casts. Hence this should be a standalone issue and be resolved in a separate PR. A synonymous `UniquePatientList` is used as interim.

**Review by jiayi1129 — COMMENTED** [review]
@Haozeee I have done the UI for risk as I did its implementation. I adapted the design from your PR.

**Comment by jiayi1129** [comment]
image
Current look for `Risk` tag UI

File: src/test/java/seedu/plannermd/model/patient/RiskTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+        // invalid addresses
+        assertFalse(Risk.isValidRisk("")); // empty string
+        assertFalse(Risk.isValidRisk(" ")); // spaces only
+
+        // valid addresses
+        assertTrue(Risk.isValidRisk("HIGH"));
+        assertTrue(Risk.isValidRisk("MEDIUM")); // one character
+        assertTrue(Risk.isValidRisk("LOW")); // long address
+    }
```
Comment (pualixue) [inline src/test/java/seedu/plannermd/model/patient/RiskTest.java:36]: Are the *addresses* supposed to be *risk*?

File: src/test/java/seedu/plannermd/storage/JsonAdaptedPatientTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    private static final String VALID_PHONE = BENSON.getPhone().toString();
+    private static final String VALID_EMAIL = BENSON.getEmail().toString();
+    private static final String VALID_ADDRESS = BENSON.getAddress().toString();
+    private static final List<JsonAdaptedTag> VALID_TAGS = BENSON.getTags().stream()
+            .map(JsonAdaptedTag::new)
+            .collect(Collectors.toList());
+    private static final String VALID_RISK = BENSON.getRisk().toString();
+
+    @Test
+    public void toModelType_validpatientDetails_returnspatient() throws Exception {
```
Comment (pualixue) [inline src/test/java/seedu/plannermd/storage/JsonAdaptedPatientTest.java:39]: Minor typo.
```suggestion
    public void toModelType_validPatientDetails_returnsPatient() throws Exception {
```

File: src/test/java/seedu/plannermd/testutil/PlannerMdBuilder.java [hunk 7]
```diff
@@ -23,8 +23,8 @@ public PlannerMdBuilder(PlannerMd plannerMd) {
     /**
      * Adds a new {@code Person} to the {@code PlannerMd} that we are building.
```
Comment (pualixue) [inline src/test/java/seedu/plannermd/testutil/PlannerMdBuilder.java:24]: Perhaps this should be Patient?
```suggestion
     * Adds a new {@code Patient} to the {@code PlannerMd} that we are building.
```

**Review by pualixue — CHANGES_REQUESTED** [review]
LGTM overall. There are a few cosmetic issues that I noticed.

File: src/test/java/seedu/plannermd/logic/commands/EditPatientCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        String expectedMessage = String.format(EditPatientCommand.MESSAGE_EDIT_PATIENT_SUCCESS, editedPatient);
+
+        Model expectedModel = new ModelManager(new PlannerMd(model.getPlannerMd()), new UserPrefs());
+        expectedModel.setPatient(model.getFilteredPatientList().get(0), editedPatient);
+
+        assertCommandSuccess(editPatientCommand, model, expectedMessage, expectedModel);
+    }
+
+    @Test
+    public void execute_duplicatePersonUnfilteredList_failure() {
```
Comment (ChuangZheQuan) [inline src/test/java/seedu/plannermd/logic/commands/EditPatientCommandTest.java:104]: maybe can change the method name to `execute_deuplicatePatientUnfilteredList_failure`?

File: src/test/java/seedu/plannermd/logic/commands/EditPatientCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        Person firstPerson = model.getFilteredPatientList().get(INDEX_FIRST_PERSON.getZeroBased());
+        EditPatientDescriptor descriptor = new EditPersonDescriptorBuilder(firstPerson).build();
+        EditPatientCommand editPatientCommand = new EditPatientCommand(INDEX_SECOND_PERSON, descriptor);
+
+        assertCommandFailure(editPatientCommand, model, EditPatientCommand.MESSAGE_DUPLICATE_PATIENT);
+    }
+
+    @Test
+    public void execute_duplicatePersonFilteredList_failure() {
```
Comment (ChuangZheQuan) [inline src/test/java/seedu/plannermd/logic/commands/EditPatientCommandTest.java:113]: same here!

File: src/test/java/seedu/plannermd/logic/commands/EditPatientCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+        assertCommandFailure(editPatientCommand, model, Messages.MESSAGE_INVALID_PATIENT_DISPLAYED_INDEX);
+    }
+
+    /**
+     * Edit filtered list where index is larger than size of filtered list,
+     * but smaller than size of plannermd
+     */
+    @Test
+    public void execute_invalidPersonIndexFilteredList_failure() {
```
Comment (ChuangZheQuan) [inline src/test/java/seedu/plannermd/logic/commands/EditPatientCommandTest.java:138]: here too!

File: src/test/java/seedu/plannermd/logic/commands/FindPatientCommandTest.java [hunk 11]
```diff
@@ -56,22 +57,22 @@ public void equals() {
 
     @Test
     public void execute_zeroKeywords_noPersonFound() {
```
Comment (ChuangZheQuan) [inline src/test/java/seedu/plannermd/logic/commands/FindPatientCommandTest.java:59]: maybe change to patient? unless can just keep it as `FindCommandTest.java` and add doctors here in the future too!

File: src/test/java/seedu/plannermd/model/patient/PatientTest.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+import seedu.plannermd.testutil.patient.PatientBuilder;
+
+class PatientTest {
+
+    @Test
+    public void equals() {
+        // same values -> returns true
+        Patient aliceCopy = new PatientBuilder(ALICE).build();
+        assertTrue(ALICE.equals(aliceCopy));
```
Comment (ChuangZheQuan) [inline src/test/java/seedu/plannermd/model/patient/PatientTest.java:24]: I'm thinking the the future we may need to differentiate the names between patients and doctors, maybe like adding the DR prefix in front of the name `DR_ALICE` for e.g., to make it clearer, wdyt?

File: src/test/java/seedu/plannermd/model/person/UniquePatientListTest.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
+package seedu.plannermd.model.person;
```
Comment (ChuangZheQuan) [inline src/test/java/seedu/plannermd/model/person/UniquePatientListTest.java:1]: maybe change the`person` in the test method names to `patient`?

**Review by ChuangZheQuan — CHANGES_REQUESTED** [review]
Overall looks good, I'm thinking maybe Person should be an abstract class, but we can discuss that in the future. Just some minor changes needed and logic looks fine. Good job!

File: docs/tutorials/TracingCode.md [hunk 14]
```diff
@@ -190,7 +190,7 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    public CommandResult execute(Model model) throws CommandException {
        ...
        Person personToEdit = lastShownList.get(index.getZeroBased());
-       Person editedPerson = createEditedPerson(personToEdit, editPersonDescriptor);
+       Person editedPerson = createEditedPerson(personToEdit, editPatientDescriptor);
```
[the remark this answers is not in the document]
Comment (jiayi1129) [reply docs/tutorials/TracingCode.md:193]: I think we need to update the whole md file together next time. So I just revert this back to `editPersonDescriptor` for now.

File: src/test/java/seedu/plannermd/model/patient/RiskTest.java [hunk 15]
```diff
@@ -22,14 +22,14 @@ public void constructor_invalidRisk_throwsIllegalArgumentException() {
...
-        // invalid addresses
+        // invalid risks
         assertFalse(Risk.isValidRisk("")); // empty string
         assertFalse(Risk.isValidRisk(" ")); // spaces only
 
-        // valid addresses
+        // valid risks
         assertTrue(Risk.isValidRisk("HIGH"));
         assertTrue(Risk.isValidRisk("MEDIUM")); // one character
         assertTrue(Risk.isValidRisk("LOW")); // long address
```
Comment (pualixue) [inline src/test/java/seedu/plannermd/model/patient/RiskTest.java]: Perhaps can remove the comments too as they are not related to risk.
```suggestion
        assertTrue(Risk.isValidRisk("MEDIUM"));
        assertTrue(Risk.isValidRisk("LOW"));
```

**Review by pualixue — APPROVED** [review]
Great job! Do consider my comment below.

File: src/main/java/seedu/plannermd/logic/commands/editcommand/EditPatientCommand.java [hunk 16]
```diff
@@ -209,18 +222,19 @@ public boolean equals(Object other) {
...
 
             return getName().equals(e.getName())
                     && getPhone().equals(e.getPhone())
                     && getEmail().equals(e.getEmail())
                     && getAddress().equals(e.getAddress())
-                    && getTags().equals(e.getTags());
+                    && getTags().equals(e.getTags())
+                    && getRisk().equals(e.getRisk());
         }
     }
```
Comment (hpkoh) [inline src/main/java/seedu/plannermd/logic/commands/editcommand/EditPatientCommand.java:239]: For our edit command to be extensible to other type of Persons, I think we could use a factory method within `EditCommand` which handles the query for the Person at `INDEX` and returns the corresponding edit command. However, there will be concerns for identity specific tokens that users may try to use with other identities.

**Review by hpkoh — CHANGES_REQUESTED** [review]
Good job on refractoring the codebase to extend to other identity for `Person`

I added some comments on approaches and philosophy we could adopt for the extension.

I have also added design philosophies zhequan, you and haoze have brought up to discuss as part of our agenda for our upcoming meeting.

**Pull request merged**